### PR TITLE
Add lazy computation of func properties

### PR DESCRIPTION
**Pull request by KPostOffice**
This likely introduces breaking changes due to change from functions being regular functions to effectively being `@property` functions.

For simple functions such as:
```
def foo(self):
    return 1
```

`@Lazy` is about twice as fast. This speedup increases as complexity increases

### CODE REVIEW

File: thoth/python/lazy.py [hunk 1]
```diff
@@ -0,0 +1,10 @@
+class Lazy(object):
```
Comment (fridex) [inline thoth/python/lazy.py:1]: Maybe suitable to `helpers.py`?

File: thoth/python/package_version.py [hunk 2]
```diff
@@ -245,10 +246,12 @@ def to_pipfile_lock(self) -> dict:
 
         return {self.name: result}
 
+    @Lazy
     def to_tuple(self) -> tuple:
         """Return a tuple representing this Python package."""
         return self.name, self.locked_version, self.index.url
 
+    @Lazy
     def to_tuple_locked(self) -> tuple:
```
Comment (fridex) [inline thoth/python/package_version.py:255]: To keep codebase consistent, what about renaming these two to property style names? e.g. `tuple` and `tuple_locked`? The implementation might live with the current `to_tuple` and `to_tuple_locked` methods until we find all the occurrences and fix them.

Comment (KPostOffice) [reply thoth/python/package_version.py:255]: I like that :+1:

**Comment by KPostOffice** [comment]
I think I can add this elsewhere in `thoth.python` WDYT?

File: thoth/python/package_version.py (hunk 2 above)
Comment (KPostOffice) [reply thoth/python/package_version.py]: I used as_tuple and as_tuple_locked because tuple is a keyword

**Comment by fridex** [comment]
> I think I can add this elsewhere in `thoth.python` WDYT?

What do you mean?

**Comment by KPostOffice** [comment]
I'm guessing there are other class functions which would benefit from this speedup?

**Comment by KPostOffice** [comment]
We need a new release of thoth common to merge this

File: thoth/python/__init__.py [hunk 3]
```diff
@@ -27,6 +27,7 @@
 from .project import Project
 from .source import Source
 from .aiosource import AIOSource, AsyncIterablePackages, AsyncIterableVersions, AsyncIterableArtifacts
+from .lazy import Lazy
```
Comment (fridex) [inline thoth/python/__init__.py:30]: Will this import work?

**Comment by KPostOffice** [comment]
I think the issue lies with the use of slots. This approach requires dynamically created properties.

**Pull request closed without merging**